Platform/OpenNtfsDxe/Compression.c

```diff
@@ -0,0 +1,668 @@
+/** @file
+  Copyright (c) 2022, Mikhail Krichanov. All rights reserved.
+  SPDX-License-Identifier: BSD-3-Clause
+
+  Functional and structural descriptions follow NTFS Documentation
+  by Richard Russon and Yuval Fledel
+**/
+
+#include "NTFS.h"
+#include "Helper.h"
+
+#define IS_COMPRESSED_BLOCK 0x8000
+#define BLOCK_LENGTH_BITS   0xFFF
+#define UNIT_MASK           0xF
+
+extern UINTN  mFileRecordSize;
+extern UINTN  mSectorSize;
+extern UINTN  mClusterSize;
+extern UINT64 BufferSize;
+
+STATIC
+EFI_STATUS
+GetNextCluster (
+  IN OUT COMPRESSED *Clusters
+  )
+{
+  EFI_STATUS   Status;
+
+  if (Clusters->Head >= Clusters->Tail) {
+    DEBUG((DEBUG_INFO, "NTFS: Compression block overflown\n"));
+    return EFI_VOLUME_CORRUPTED;
+  }
+
+  Status = DiskRead (
+    Clusters->FileSystem,
+    (Clusters->Elements[Clusters->Head].Lcn - Clusters->Elements[Clusters->Head].Vcn + Clusters->CurrentVcn) * mClusterSize,
+    mClusterSize,
+    Clusters->Cluster
+    );
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  Clusters->CurrentVcn++;
+
+  if (Clusters->CurrentVcn >= Clusters->Elements[Clusters->Head].Vcn) {
+    Clusters->Head++;
+  }
+
+  Clusters->ClusterOffset = 0;
+
+  return EFI_SUCCESS;
+}
+
+STATIC
+EFI_STATUS
+GetDataRunByte (
+  IN  COMPRESSED *Clusters,
+  OUT UINT8      *Result
+  )
+{
+  EFI_STATUS   Status;
+
+  if (Clusters->ClusterOffset >= mClusterSize) {
+    Status = GetNextCluster (Clusters);
+    if (EFI_ERROR (Status)) {
+      return Status;
+    }
+  }
+
+  *Result = Clusters->Cluster[Clusters->ClusterOffset++];
+
+  return EFI_SUCCESS;
+}
+
+STATIC
+EFI_STATUS
+GetTwoDataRunBytes (
+  IN  COMPRESSED *Clusters,
+  OUT UINT16     *Result
+  )
+{
+  EFI_STATUS   Status;
+  UINT8        ByteLow  = 0;
+  UINT8        ByteHigh = 0;
+
+  Status = GetDataRunByte (Clusters, &ByteLow);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  Status = GetDataRunByte (Clusters, &ByteHigh);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  *Result = (((UINT16) ByteHigh) << 8) + ByteLow;
+
+  return EFI_SUCCESS;
+}
+
+/**
```

Comment:
The documentation style here looks strange to me. Perhaps take a look at how others are written?

Reply:
The purpose of my documentation is to take an every useful thing from a piece of sh*t, called NTFS documentation, so as other programmers would not have to waste their time on reading such an abomination several times from the very beginning to the very end to make a head or tail of what is going on in this driver. Current style of comments make them foldable in my text editor but you are welcome to give ideas, which will make this driver satisfy OpenCore style.